observer/probers/tls/tls.go

```diff
@@ -0,0 +1,165 @@
+	"crypto/x509"
+	"encoding/base64"
+	"errors"
+	"fmt"
+	"github.com/prometheus/client_golang/prometheus"
+	"golang.org/x/crypto/ocsp"
+	"io"
+	"net/http"
+	"time"
+)
+
+var StatusCode int
+
+const (
+	Success = iota
+	UnexpectedRoot
+	UnexpectedResponse
+	UnexpectedRootAndResponse
+	UnknownOCSPCheck
+	FailedOCSPCheck
+	FailedTLSConnection
+	Unknown
+)
+
+// TLSProbe is the exported 'Prober' object for monitors configured to
+// perform TLS protocols.
+type TLSProbe struct {
+	url        string
+	root       string
+	response   string
+	certExpiry *prometheus.GaugeVec
+	StatusCode *prometheus.GaugeVec
+}
+
+// Name returns a string that uniquely identifies the monitor.
+func (p TLSProbe) Name() string {
+	return fmt.Sprintf("%s--expecting-%s-%s", p.url, p.response, p.root)
+}
+
+// Kind returns a name that uniquely identifies the `Kind` of `Prober`.
+func (p TLSProbe) Kind() string {
+	return "TLS"
+}
+
+func isOCSPRevoked(cert, issuer *x509.Certificate) bool {
+	req, err := ocsp.CreateRequest(cert, issuer, nil)
+	if err != nil {
+		StatusCode = FailedOCSPCheck
+		return false
+	}
+
+	url := fmt.Sprintf("%s/%s", cert.OCSPServer[0], base64.StdEncoding.EncodeToString(req))
+	res, err := http.Get(url)
+	if err != nil {
+		StatusCode = FailedOCSPCheck
+		return false
+	}
+
+	output, err := io.ReadAll(res.Body)
+	if err != nil {
+		StatusCode = FailedOCSPCheck
+		return false
+	}
+
+	ocspRes, err := ocsp.ParseResponseForCert(output, cert, issuer)
+	if err != nil {
+		StatusCode = FailedOCSPCheck
+		return false
+	}
+
+	if ocspRes.Status == ocsp.Revoked {
+		return true
+	} else {
+		if ocspRes.Status == ocsp.Unknown {
+			StatusCode = UnknownOCSPCheck
+		}
+		return false
+	}
+}
+
+func (p TLSProbe) getExpiredCertInfo() (string, string, time.Duration) {
+	conf := &tls.Config{InsecureSkipVerify: true}
+	conn, err := tls.Dial("tcp", p.url+":443", conf)
+	if err != nil {
+		StatusCode = FailedTLSConnection
+		return "", "", 0
+	}
+	peers := conn.ConnectionState().PeerCertificates
+	expiry_date := peers[0].NotAfter
+	root := peers[len(peers)-1].Issuer
+	return root.CommonName, root.Organization[0], time.Since(expiry_date)
+}
+
+func updateStatusCode(is_expected_root, is_expected_response bool) {
+	if !is_expected_root && !is_expected_response {
+		StatusCode = UnexpectedRootAndResponse
+	} else if !is_expected_root {
+		StatusCode = UnexpectedRoot
+	} else if !is_expected_response {
+		StatusCode = UnexpectedResponse
+	}
+}
+
+// Probe performs the configured TLS protocol.
+// Return true if both root AND response are the expected values, otherwise false
+// Export time to cert expiry as Prometheus metric
+func (p TLSProbe) Probe(timeout time.Duration) (bool, time.Duration) {
+	is_expected_root, is_expected_response := false, false
+	root_cn, root_o := "", ""
+	var time_to_expiry time.Duration
+	var time_since_expiry time.Duration
+	StatusCode = Success
+
+	conf := &tls.Config{}
+	start := time.Now()
+	conn, err := tls.Dial("tcp", p.url+":443", conf)
```

Comment:
This can just be inlined :)

```suggestion
	start := time.Now()
	conn, err := tls.Dial("tcp", p.url+":443", &tls.Config{})
```